x/cosmwasmpool/README.md

```diff
@@ -0,0 +1,284 @@
+
+The module is built on top of the CosmWasm smart contracting platform, which provides a secure and efficient way to develop and execute WebAssembly (Wasm) smart contracts on the Cosmos SDK.
+
+Having pools in CosmWasm provides several benefits, one of which is avoiding the need for chain upgrades when introducing new functionalities or modifying existing ones related to liquidity pools. This advantage is particularly important in the context of speed of development and iteration.
+
+An example of a CosmWasm pool type: [transmuter](https://github.com/osmosis-labs/transmuter)
+
+
+## Creating new CosmWasm Pool
+
+To create new CosmWasm Pool, there are 3 modules involved: `x/cosmwasmpool`, `x/wasm`, and `x/poolmanager`. Here is an overview of the process:
+
+```mermaid
+graph TD;
+  Sender((Sender))
+
+  Sender -- create poool --> x/cosmwasmpool
+  x/cosmwasmpool -- get next & set pool id --> x/poolmanager
+  x/cosmwasmpool -- instantiate contract --> x/wasm
+```
+
+The CosmWasm contract that is to be instanitiated needs to implement [CosmWasm Pool Contract Interface](#cosmwasm-pool-contract-interface) and store it on chain first. Then new pool can be created by sending `MsgCreateCosmWasmPool`.
+
+
+`MsgCreateCosmWasmPool` contains `InstantiateMsg`, which is a message that will be passed to the CosmWasm contract when it is instantiated. The structure of the message is defined by the contract developer, and can contain any information that the contract needs to be instantiated. JSON format is used for `InstantiateMsg`.
+
+```mermaid
+sequenceDiagram
+    participant Sender
+    participant x/cosmwasmpool
+    participant x/wasm
+    participant x/poolmanager
+
+    Sender ->> x/cosmwasmpool: MsgCreateCosmWasmPool {CodeId, InstantiateMsg, Sender}
+
+    Note over x/wasm: Given there is a pool contract with CodeId
+
+
+    x/cosmwasmpool ->> x/wasm: Call InstantiateContract(CodeId, InstantiateMsg)
+    x/wasm -->> x/cosmwasmpool: ContractAddress
+    x/poolmanager ->> x/cosmwasmpool: Call GetNextPoolId()
+    x/cosmwasmpool ->> x/poolmanager: Call SetNextPoolId(poolId)
+
+    Note over x/cosmwasmpool: Store CodeId, ContractAddress, and PoolId
+
+    x/cosmwasmpool -->>  Sender: MsgCreateCosmWasmPoolResponse {PoolId}
+```
+
+
+## Providing / Withdrawing Liquidity
+
+Currently, all existing pool types has their own way of providing liquidity and shares calculation. CosmWasm pool aims to be flexible that regards and let the contract define the way of providing liquidity. So there is no restriction here, and the contract developer can define the way of providing liquidity as they wish, potentially with execute endpoint since `MsgExecuteContract` triggers state mutating endpoint and can also attach funds to it.
```

Comment:
```suggestion
Currently, all existing pool types have their own way of providing liquidity and shares calculation. CosmWasm pool aims to be flexible that regards and let the contract define the way of providing liquidity. So there is no restriction here, and the contract developer can define the way of providing liquidity as they wish, potentially with execute endpoint since `MsgExecuteContract` triggers state mutating endpoint and can also attach funds to it.
```